### DISCUSSION TITLE
Pyright Misinterprets DataFrame Operation Output

### DISCUSSION THREAD

**Discussion by GhostlyCipher** [microsoft/pyright#7636] [Q&A] [answered]
Does this code work? Yes.
Only type checking is failing.

I have a pandas DataFrame where I'm doing these operations.

```
import pandas as pd

def cleaup(df: pd.DataFrame) -> pd.DataFrame:
    df['col_name_1'] = df['col_name_1'].astype('string')
    df = df.drop_duplicates(subset=['col_name_2'])
    return df
```

`df.drop_duplicates` is supposed to return a `DataFrame`

Function definition is:
```
(method) def drop_duplicates(
    subset: ... = ...,
    *,
    keep: NaPosition | _bool = ...,
    inplace: _bool = ...,
    ignore_index: _bool = ...
) -> DataFrame
```

Error
```
error: Expression of type "DataFrame | None" cannot be assigned to return type "DataFrame"
    Type "DataFrame | None" cannot be assigned to type "DataFrame"
      "None" is incompatible with "DataFrame" (reportReturnType)
```

`df` gets a type of `DataFrame | None` by pyright while it should be `DataFrame`.

**Comment by erictraut** [comment]
This is unlikely to be a bug in pyright and more likely an issue with the pandas stubs. I may be able to help confirm if you provide a minimal, self-contained code sample and repro steps. Your code sample above references symbols that are undefined, so it is incomplete.

**Comment by GhostlyCipher** [comment]
Here is a self contained example which is raising a couple of errors.

```
import pandas as pd

# Define the cleanup function
def cleanup(df: pd.DataFrame) -> pd.DataFrame:
    df['col_name_1'] = df['col_name_1'].astype('string')
    df = df.drop_duplicates(subset=['col_name_2'])
    return df

# Create a sample DataFrame
data = {'col_name_1': [1, 2, 3, 4, 5],
        'col_name_2': ['A', 'B', 'C', 'A', 'D']}
df = pd.DataFrame(data)
cleanup(df)
```

Errors:
```
test-pyright.py:5:41 - error: "astype" is not a known member of "None" (reportOptionalMemberAccess)
test-pyright.py:6:10 - error: Expression of type "DataFrame | None" cannot be assigned to declared type "DataFrame"
```

**Reply by Azureblade3808** [reply, marked as answer]
I highly suspect that you are using a bare Pandas without corresponding stubs. If so, take a look at https://github.com/pandas-dev/pandas-stubs .

**Reply by GhostlyCipher** [reply]
Thanks for pointing this out. I didn't have pandas-stubs installed.
